### ISSUE TITLE
metaclass

### ISSUE THREAD

**Issue by allyourcode** [micropython/micropython#6094]
I see that MP does not (yet) support metaclasses (per item #6 on https://github.com/micropython/micropython/wiki/Differences#differences-by-design). It sounds like the MP developers would like to add this feature some day, but it's not clear to me how strongly they feel about that.

Obviously, people do not directly make use of this feature very often. Nevertheless, it is very powerful, and people DO benefit from libraries that do use it directly. I am working on a library that works in CPython, but because it hinges on a metaclass, I was disappointed to find that it doesn't work on MP.

My hopes in posting this are to

a) Solicit the thoughts of the MP developers on where this feature lies in their priority queue.
b) Let others provide "me too!" feedback.
c) Vent my frustration.

PS: Thank you for MP! Being able to use (u)asyncio is truly a godsend.

**Comment by dpgeorge** [comment]
> Obviously, people do not directly make use of this feature very often. Nevertheless, it is very powerful, and people DO benefit from libraries that do use it directly.

Yes, that's also my understanding, that metaclasses are not really a user-facing feature but rather useful to implement things that are user facing.

> Solicit the thoughts of the MP developers on where this feature lies in their priority queue.

For myself, it's very low priority, and we've not really had any requests for it in the past.  The main deciding factor for features like this is not so much how hard it is to add, but rather how much it makes the core interpreter grow in terms of size and complexity.

A good first task would be to write a few small test cases which show how metaclasses should behave.

**Comment by winni2k** [comment]
I'm looking at porting dataclasses to MP, and I see that the CPython 3.8 implementation of dataclasses relies on metaclasses (https://github.com/python/cpython/blob/4d58a1d8fb27048c11bcbda3da1bebf78f979335/Lib/dataclasses.py#L206).

**Comment by EytanDn** [comment]
Hi, MP 1.26.0 user here, are there any future plans on this? metaclasses would be indeed very powerful tool

**Comment by Josverl** [comment]
What is the use case you need metaclasses  for?

While it is definitely useful,
most use cases can be implemented as well using decorators or inheritance.

https://realpython.com/python-metaclasses/#is-this-really-necessary

**Comment by bartek-w** [comment]
I believe this is mostly about easy code portability. It would be great to be able to easily implement a fully functional Enum in MicroPython directly (or similar classes). Maybe MicroPython could support some basic metaclass functionality or add support for __init_subclass__, which would allow implementing substitutions for most basic metaclass functionality?

The metaclasses support is the main blocker when I want to easily copy&paste existing python code and get it running. What are the main blockers to add some metaclass functionality? Added implementation complexity? Memory or CPU overhead? Or metaclasses are simply too much for Micropython from the project design assumptions, as it should be kept 'micro'?

**Comment by Josverl** [comment]
I drafted a Q&D prototype to get a feel for the cost  - note that I did not yet look at runtime performance overhead.

Component | 	Initial Cost| 	Optimized Cost |	
-----|-----|-----|
metaclass Keyword extraction & **kwargs support | 	~200 bytes | 	~50 bytes | 	~150 bytes
`__prepare__` method support | 	~250-300 bytes | 	0 bytes (optional) | 	~250-300 bytes
Metaclass resolution algorithm | 	~250-300 bytes | 	~150-200 bytes | 	
Argument array building | 	~150-200 bytes | 	~50 bytes | 	 

So the size impact is somewhere between  0 - 150 - 900 bytes , depending on the CPython compatibility desired.
No integral testing of runtime overhead/slowdown has been taken into account. 

I think the next relevant Question is which sub-features are needed most  to implement things like:

- **Enum** 
  - https://github.com/micropython/micropython-lib/issues/269
  - https://github.com/micropython/micropython-lib/pull/980
  - https://github.com/micropython/micropython/pull/16842
- **DataClass**
  - https://github.com/orgs/micropython/discussions/13741
  - https://github.com/orgs/micropython/discussions/9327
- [User defined Generics](https://typing.python.org/en/latest/spec/generics.html#user-defined-generic-types) (typing)
  - https://github.com/micropython/micropython-lib/pull/584
  - https://github.com/Josverl/micropython-stubs/issues/806

~~Q&D-PR : https://github.com/Josverl/micropython/pull/25~~
[WIP PR](https://github.com/micropython/micropython/pull/18362)

Optimizations :
- Guarded by MICROPY_METACLASS (default TBD) 
 - Made `__prepare__` optional - New MICROPY_METACLASS_PREPARE flag (default: 0)
- Simplified keyword handling - Only metaclass= keyword supported, other **kwargs ignored
- Optimized resolution - Reduced validation complexity while maintaining conflict detection
- Removed alloca() - Use fixed-size stack arrays

Trade-offs (documented in cpydiff tests)
- `__prepare__ `method not called unless MICROPY_METACLASS_PREPARE=1
- Additional **kwargs like class C(metaclass=M, foo='bar'): are ignored
- Core functionality retained (metaclass=, inheritance, init, conflicts)

~~~Size cost : 152 bytes (+0.019%) vs baseline Unix port ~~

**Comment by EytanDn** [comment]
> What is the use case you need metaclasses for?
> 
> While it is definitely useful, most use cases can be implemented as well using decorators or inheritance.
> 
> https://realpython.com/python-metaclasses/#is-this-really-necessary

I intended on making a metaclass for singleton objects which I need to keep track of their properties in RTC memory, kinda similar to arduino RTC_DATA_ATTR, firstly, singleton meta class works alot better in my experience than other python options for singletons, and secondly , I get that I can do that with an property.setter decorator but It would be much easier to simply define the metaclass to take care of that, then I can more freely mess around with my objects properties, adding/removing properties with cleaner code

**Comment by Josverl** [comment]
@EytanDn - still work to be done, but would be interested to see if https://github.com/micropython/micropython/pull/18362 would help for your use cases.
